src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameHeaders.cs

```diff
@@ -233,16 +233,57 @@ public static void ValidateHeaderCharacters(string headerCharacters)
             }
         }
 
-        public static long ParseContentLength(StringValues value)
+        public static unsafe long ParseContentLength(StringValues value)
         {
-            try
-            {
-                return long.Parse(value, NumberStyles.AllowLeadingWhite | NumberStyles.AllowTrailingWhite, CultureInfo.InvariantCulture);
-            }
-            catch (FormatException ex)
+            const string errorMessage = "Content-Length value must be an integral number.";
+            var input = value.ToString();
+            var parsed = (long)0;
+
+            fixed (char* ptr = input)
             {
-                throw new InvalidOperationException("Content-Length value must be an integral number.", ex);
+                var ch = ptr;
+                var end = ptr + input.Length;
+
+                // Skip leading whitespace
+                while (ch < end && ((*ch >= 0x09 && *ch <= 0x0D) || *ch == 0x20))
+                {
+                    ch++;
+                }
+
+                // If done on end of input or char is not a number, input is invalid
+                if (ch == end || *ch < 0x30 || *ch > 0x39)
+                {
+                    throw new InvalidOperationException(errorMessage);
```

Comment:
Don't throw inline; move out out a separate method e.g. `void ThrowInvalidContentLength()` else it will bring in 3 `InvalidOperationException` ctors and 3 string loads